xarray/test/test_dataset.py

```diff
@@ -2916,6 +2917,44 @@ def test_filter_by_attrs(self):
         for var in new_ds.data_vars:
             self.assertEqual(new_ds[var].height, '10 m')
 
+    def test_binary_op_join_setting(self):
+        """
+        A test method to verify the ability to set binary operation join kwarg
+        ("inner", "outer", "left", "right") via xr.set_options().
+        """
+        # First we set up a data array
+        xdim, ydim, zdim = 'x', 'y', 'z'
+        xcoords, ycoords, zcoords = ['a', 'b', 'c'], [-2, 0, 2], [0, 1, 2]
+        total_size = len(xcoords) * len(ycoords) * len(zcoords)
+        # create a 3-by-3-by-3 data array
+        arr = DataArray(np.arange(total_size).\
+                        reshape(len(xcoords),len(ycoords),len(zcoords)),
+                        [(xdim, xcoords), (ydim, ycoords),(zdim, zcoords)])
+        # now create a data array with the last x slice missing
+        arr1 = arr[0:-1,:,:].copy()
+        missing_last_x = arr1.to_dataset(name='foo')
+        # create another data array with the last z slice missing
+        arr2 = arr[:,:,0:-1].copy()
+        missing_last_z = arr2.to_dataset(name='foo') # needs to be name='foo' as well
+        # because the default in OPTIONS is join="inner", we test "outer" first
+        xr.set_options(join="outer")
+        result = missing_last_x + missing_last_z
+        self.assertTrue(result.foo.shape == arr.shape)
+        self.assertTrue(result.foo[-1,:,:].isnull().all())
+        self.assertTrue(result.foo[:,:,-1].isnull().all())
+        # now revert back to join="inner"
+        xr.set_options(join="inner")
+        result = missing_last_x + missing_last_z
+        self.assertTrue(result.foo.shape == \
+                        (len(xcoords)-1, len(ycoords), len(zcoords)-1))
+        self.assertTrue(result.foo.notnull().all())
+        self.assertFalse('c' in list(result.foo['x']))
+        self.assertFalse(2 in list(result.foo['z']))
+        # just for kicks, what happens when the dataarrays have different names?
```

Comment:
We already have tests for similar behavior -- see `test_dataset_math_auto_align`. This might belong there (more tests are great!), but it doesn't belong here.

It does raise the interesting question of what the behavior should be when `join='outer'` for Dataset objects with different data variables. Should the result now have the union of data variable names? If so, this will take a bit more retooling. (This would be easier after https://github.com/pydata/xarray/pull/964 is merged)